### PR TITLE
Enforce nonconvex for OffsetConverter on a single flow instead of outputs

### LINKED ISSUE #1010: Misleading/missing documentation of OffsetConverter
**Describe the bug**
In the [usage documentation for the OffsetTransformer](https://oemof-solph.readthedocs.io/en/stable/usage.html#offsetconverter-component) an example is given to calculate the coefficients `c0` and `c1`. If calculated like this, the results are illogical. I narrowed it down to the calculation of `c0`, which should actually  look like this: 
`c0 = (P_out_max-c1*P_in_max)/P_out_max`
Which means, that `c0` isn't really the y-intercept anymore (which would be `c0 = P_out_max-c1*P_in_max`).

**Possible solutions**
I see two possible solutions:
1) Fix the documentation to include the example calculations of the parameters described above, or
2) Change the calculation in the `OffsetConverterBlock` to not multiply with the `status_nominal` (which is the product of `P_max(t) * Y(t)`) but with the `status` variable. This way, `c0` would remain to be the y-intercept, which, from a mathematical, logical and usage standpoint seems much nicer to me.

Am I overlooking something here or doing something wrong? I'm actually a bit surprised that this has not been noticed before.

### LINKED ISSUE #1067: The `OffsetConverter` is broken in the new implementation
**Describe the bug**
With the current formulation, the `OffsetConverter` can produce the outputs at the level of each respective `offset` without consuming anything from the input.

I am preparing a PR to solve the issue.

**To Reproduce**
Consider the following example of an electrolyzer producing hydrogen and heat. The electrolyzer follows a fixed hydrogen demand, a slack source is available in case the demand cannot be supplied due to part load restrictions.

```python
from oemof import solph


es = solph.EnergySystem(timeindex=solph.create_time_index(year=2023, number=5))

eta_h2_min = 0.6                 # efficiency at minimal operation point
eta_h2_max = 0.5                 # efficiency at nominal operation point
P_in_min = 20                   # absolute minimal output power
P_in_max = 100                  # absolute nominal output power

# calculate limits of input power flow
P_out_min_h2 = P_in_min * eta_h2_min
P_out_max_h2 = P_in_max * eta_h2_max

# calculate coefficients of input-output line equation
slope_h2 = (P_out_max_h2 - P_out_min_h2) / (P_in_max - P_in_min)
offset_h2 = (P_out_max_h2 - slope_h2 * P_in_max) / P_out_max_h2


eta_heat_min = 0.3                 # efficiency at minimal operation point
eta_heat_max = 0.4                 # efficiency at nominal operation point

# calculate limits of input power flow
P_out_min = P_in_min * eta_heat_min
P_out_max = P_in_max * eta_heat_max

# calculate coefficients of input-output line equation
slope_heat = (P_out_max - P_out_min) / (P_in_max - P_in_min)
offset_heat = (P_out_max - slope_heat * P_in_max) / P_out_max

b_h2 = solph.Bus("hydrogen bus")
b_heat = solph.Bus("heat bus")
b_el = solph.Bus("electricity bus")

b_electrolyzer_h2 = solph.Bus("electrolyzer virtual input for h2")
b_electrolyzer_heat = solph.Bus("electrolyzer virtual input for heat")

source_el = solph.components.Source("electricity import", outputs={b_el: solph.Flow()})
source_slack_h2 = solph.components.Source("hydrogen slack", outputs={b_h2: solph.Flow(variable_costs=1000)})
sink_heat = solph.components.Sink("heat export", inputs={b_heat: solph.Flow()})
sink_h2 = solph.components.Sink("hydrogen export", inputs={b_h2: solph.Flow(fix=[10, 20, 30, 40, 50, 60], nominal_value=1)})

electrolyzer = solph.components.OffsetConverter(
    label="electrolyzer",
    inputs={b_el: solph.Flow(nominal_value=P_in_max)},
    outputs={
        b_heat: solph.Flow(
            nominal_value=P_out_max,
            nonconvex=solph.NonConvex(),
            min=P_out_min / P_out_max
        ),
        b_h2: solph.Flow(
            nominal_value=P_out_max_h2,
            nonconvex=solph.NonConvex()
        )
    },
    coefficients={
        b_heat: [offset_heat, slope_heat],
        b_h2: [offset_h2, slope_h2],
    }
)

es.add(
    b_el, b_heat, b_h2,
    source_el, sink_h2, sink_heat, source_slack_h2,
    electrolyzer
)

om = solph.Model(es)

om.solve("gurobi")

result = solph.views.convert_keys_to_strings(om.results())
result[("electrolyzer", "hydrogen bus")]["sequences"]
```
The minimum hydrogen production should be at `12`, but due to the offset it is at `2.5`.

|                     |   flow |   status |   status_nominal |
|:--------------------|-------:|---------:|-----------------:|
| 2023-01-01 00:00:00 |    2.5 |        1 |               50 
| 2023-01-01 01:00:00 |   20   |        1 |               50 |
| 2023-01-01 02:00:00 |   30   |        1 |               50 |
| 2023-01-01 03:00:00 |   40   |        1 |               50 |
| 2023-01-01 04:00:00 |   50   |        1 |               50 |
| 2023-01-01 05:00:00 |   50   |        1 |               50 |

The reason for that is, that the `status_nominal` is multiplied with the output binary variable instead of the input binary variable.

### PR DESCRIPTION
**Pull request by fwitte**
Resolve #1010

### CODE REVIEW

**Comment by p-snft** [comment]
Thank you. I think this issue came up as the offset is typically negative, so there would be no output without input. However, when taking into account the losses (as you did in #1067) you will always run into the issue. So, I'd also consider the current implementation broken.

As you suggested in the chat, it might make sense to change the keyword argument, when we change the implementation anyway. I'd suggest `conversion_factors` for the slope, as it's the same as for the (simple) `Converter`, and `normed_offset` to emphasise the fact that it's normed.

**Comment by fwitte** [comment]
> Thank you. I think this issue came up as the offset is typically negative, so there would be no output without input. However, when taking into account the losses (as you did in #1067) you will always run into the issue. So, I'd also consider the current implementation broken.

Agree!

> As you suggested in the chat, it might make sense to change the keyword argument, when we change the implementation anyway. I'd suggest `conversion_factors` for the slope, as it's the same as for the (simple) `Converter`, and `normed_offset` to emphasise the fact that it's normed.

That sounds like a good idea. I will complete the changes in this PR then. This includes

- rename the keywords
- update the docs and examples
- fix/create tests

**Comment by fwitte** [comment]
On top of that, I suggest a couple more changes:

- Rename to `MultiOutputOffsetConverter`, and we could add a new component called `MultiInputOffsetConverter` or
- enforce a single flow of all inputs and outputs to be `NonConvex`. This flow would be the basis for all other `conversion_factors` and `normed_offsets`. Therefore it also provides the `min` and `max` boundaries, as well as the `status_nominal` and also the `Invest` if it should be considered.

The second version would be more simple from the user perspective (you have to define a reference flow, i.e. the `NonConvex` one, and all given offsets and slopes refer to that, no matter if it has multi input, multi output or both, and no matter of that `NonConvex` flow is an input or an output. The other version is likely more easy to implement (or maintain), but I am not really sure. What do you think?

**Comment by fwitte** [comment]
>     * enforce a single flow of all inputs and outputs to be `NonConvex`. This flow would be the basis for all other `conversion_factors` and `normed_offsets`. Therefore it also provides the `min` and `max` boundaries, as well as the `status_nominal` and also the `Invest` if it should be considered.

I drafted this version with my latest commit. I will readd the investment part after I have created two examples which we should discuss before putting more time into the changes.

**Comment by fwitte** [comment]
## Example 1

electricity input with hydrogen (efficiency declining with load) and heat (efficiency increasing with load) output, reference is the only input flow:

offset-two-output

```python
from oemof import solph
import numpy as np


t_steps = 11
es = solph.EnergySystem(timeindex=solph.create_time_index(year=2023, number=t_steps - 1))


eta_h2_min = 0.6
eta_h2_max = 0.5
P_in_min = 2
P_in_max = 10

P_out_min_h2 = P_in_min * eta_h2_min
P_out_max_h2 = P_in_max * eta_h2_max

slope_h2 = (P_out_max_h2 - P_out_min_h2) / (P_in_max - P_in_min)
offset_h2 = (P_out_max_h2 - slope_h2 * P_in_max) / P_in_max


eta_heat_min = 0.25
eta_heat_max = 0.3
P_out_min_heat = P_in_min * eta_heat_min
P_out_max_heat = P_in_max * eta_heat_max

slope_heat = (P_out_max_heat - P_out_min_heat) / (P_in_max - P_in_min)
offset_heat = (P_out_max_heat - slope_heat * P_in_max) / P_in_max



b_h2 = solph.Bus("hydrogen bus")
b_heat = solph.Bus("heat bus")
b_el = solph.Bus("electricity bus")

b_electrolyzer_h2 = solph.Bus("electrolyzer virtual input for h2")
b_electrolyzer_heat = solph.Bus("electrolyzer virtual input for heat")

source_el = solph.components.Source("electricity import", outputs={b_el: solph.Flow()})
sink_heat = solph.components.Sink("heat export", inputs={b_heat: solph.Flow()})
sink_h2 = solph.components.Sink("hydrogen export", inputs={b_h2: solph.Flow(fix=[0] + np.linspace(P_out_min_h2, P_out_max_h2, t_steps - 1).tolist(), nominal_value=1)})

electrolyzer = solph.components.OffsetConverter(
    label="electrolyzer",
    inputs={b_el: solph.Flow(nominal_value=P_in_max, nonconvex=solph.NonConvex(), min=P_in_min / P_in_max)},
    outputs={
        b_heat: solph.Flow(),
        b_h2: solph.Flow()
    },
    conversion_factors={
        b_heat: slope_heat,
        b_h2: slope_h2,
    },
    normed_offsets={b_heat: offset_heat, b_h2: offset_h2}
)

es.add(
    b_el, b_heat, b_h2,
    source_el, sink_h2, sink_heat,
    electrolyzer
)

om = solph.Model(es)

om.solve("gurobi")

result = solph.views.convert_keys_to_strings(om.results())
```

```python
heat = result[("electrolyzer", "heat bus")]["sequences"]["flow"]
el = result[("electricity bus", "electrolyzer")]["sequences"]["flow"]
h2 = result[("electrolyzer", "hydrogen bus")]["sequences"]["flow"]

plot_lines_and_conversion_efficiencies(
    {"flows": el, "min": P_in_min, "max": P_in_max, "commodity": "electricity input"},
    {"flows": heat, "offset": offset_heat, "slope": slope_heat, "commodity": "heat output"},
    {"flows": h2, "offset": offset_h2, "slope": slope_h2, "commodity": "h2 output"},
)
```

## Example 2

gas input and electricity output with electricity **(output commodity)** as reference, might not make that much sense, but to show, that it works.

offset-single-output

```python
from oemof import solph
import numpy as np


t_steps = 11
es = solph.EnergySystem(timeindex=solph.create_time_index(year=2023, number=t_steps - 1))


eta_el_min = 0.3
eta_el_max = 0.4
P_in_min = 2
P_in_max = 10

P_out_min_el = P_in_min * eta_el_min
P_out_max_el = P_in_max * eta_el_max

slope_gas = (P_in_max - P_in_min) / (P_out_max_el - P_out_min_el)
offset_gas = (P_in_max - slope_gas * P_out_max_el) / P_out_max_el


b_el = solph.Bus("electricity bus")
b_gas = solph.Bus("gas bus")

source_gas = solph.components.Source("gas import", outputs={b_gas: solph.Flow()})
sink_el = solph.components.Sink("electricity export", inputs={b_el: solph.Flow(fix=[0] + np.linspace(P_out_min_el, P_out_max_el, t_steps - 1).tolist(), nominal_value=1)})

chp = solph.components.OffsetConverter(
    label="chp",
    inputs={b_gas: solph.Flow()},
    outputs={
        b_el: solph.Flow(nominal_value=P_out_max_el, nonconvex=solph.NonConvex(), min=P_out_min_el / P_out_max_el)
    },
    conversion_factors={b_gas: slope_gas},
    normed_offsets={b_gas: offset_gas}
)

es.add(
    b_gas, b_el,
    source_gas, sink_el,
    chp
)

om = solph.Model(es)

om.solve("gurobi")

result = solph.views.convert_keys_to_strings(om.results())

el = result[("chp", "electricity bus")]["sequences"]["flow"]
gas = result[("gas bus", "chp")]["sequences"]["flow"]

plot_lines_and_conversion_efficiencies(
    {"flows": el, "min": P_out_min_el, "max": P_out_max_el, "commodity": "electricity output"},
    {"flows": gas, "offset": offset_gas, "slope": slope_gas, "commodity": "gas input"},
)
```

## Plotting code

```pythonimport matplotlib.pyplot as plt


def plot_lines_and_conversion_efficiencies(x, *args):

    fig, ax = plt.subplots(2, len(args), sharex=True)

    x_min = x["min"]
    x_max = x["max"]
    x_commodity = x["commodity"]
    x = x["flows"]

    ax = ax.transpose().flatten()

    for i, y in enumerate(args):
        offset = y["offset"]
        slope = y["slope"]
        y_commodity = y["commodity"]
        y = y["flows"]

        ax[2 * i].plot([0, x_min], [offset * x_max, offset * x_max + slope * x_min], "--", color="k")
        ax[2 * i].plot([x_min, x_max], [offset * x_max + slope * x_min, (offset + slope) * x_max], color="k")
        ax[2 * i].scatter(x, y)

        ax[2 * i].set_ylabel(y_commodity)
        ax[2 * i].grid()

        ax[2 * i + 1].scatter(x, y / x)
        ax[2 * i + 1].set_ylabel(f"{y_commodity} to {x_commodity} ratio")
        ax[2 * i + 1].grid()
        ax[2 * i + 1].set_xlabel(x_commodity)

    plt.tight_layout()
    fig.savefig("figure.png")
```

**Comment by fwitte** [comment]
I added two "convenience functions", i.e. `calculate_slope_and_offset_with_reference_to_ouput` and `calculate_slope_and_offset_with_reference_to_input`. You provide the output/input nominal power, minimal power as well as the efficiencies at both points and it will give you the slope and the offset to describe those points. I think, it might be helpful for the users to have them available within `oemof-solph`. On the other hand, they somehow work like a preprocessing step for the `OffsetConverter`, therefore they could also be integrated as an own component masquerading/facading the `OffsetConverter` as alternative.

What do you think, is there a place for this kind of function inside `oemof-solph`?

**Review by Bachibouzouk — COMMENTED** [review]
Thanks @fwitte for this PR :) Nice work!

The review I made is not very in depth, I haven't tested it locally because of a lack of example I could readily run (i.e. lack of time to build such an example). However I still have a few suggestions that might be relevant:
- an example like https://github.com/oemof/oemof-solph/blob/dev/examples/offset_converter_example/offset_diesel_genset_nonconvex_investment.py would be nice
- what about the back compatibility? Shouldn't this component bear another name to avoid breaking changes? 
- an amendment to the documentation of the OffsetConverter would be good

File: docs/usage.rst [hunk 1]
```diff
@@ -688,31 +688,43 @@ The following example illustrates how to define an OffsetConverter for given inf
...
-    # calculate coefficients of input-output line equation
-    c1 = (l_max-l_min)/(l_max/eta_max - l_min/eta_min)
-    c0 = l_min * (1-c1/eta_min)
+    >>> from oemof import solph
+
+    >>> eta_min = 0.5                # efficiency at minimal operation point
+    >>> eta_max = 0.8                # efficiency at nominal operation point
+    >>> P_out_min = 20               # absolute minimal output power
+    >>> P_out_max = 100              # absolute nominal output power
+    >>> l_max = P_out_max/P_out_max  # upper part load limit
```
[the remark this answers is not in the document]
Comment (fwitte) [reply docs/usage.rst:697]: `l_max` in this case should actually be 1 (it was that way in the old version of the docs already, I did not change those lines except make them doc-testable). I think the only reasonable place for it not being 1, would be, if the min and max are time series or something like that. I could also remove that line and just pass 1 in the respective places...

What do you think?

[the remark this answers is not in the document]
Comment (Bachibouzouk) [reply docs/usage.rst:697]: I think it is better to remain explicit then and thus not remove this line and simply set 1. I would be interested to know more about  the usecases where it is not 1

File: src/oemof/solph/components/_offset_converter.py [hunk 2]
```diff
@@ -72,111 +74,147 @@ class OffsetConverter(Node):
...
+            for flow in outputs.values():
+                max = flow.max[0]
+                min = flow.min[0]
+                eta_at_max = max * coefficients[1] / (max - coefficients[0])
+                eta_at_min = min * coefficients[1] / (min - coefficients[0])
+
+                slope, offset = calculate_slope_and_offset_with_reference_to_output(max, min, eta_at_max, eta_at_min)
+                conversion_factors = {input_bus: slope}
+                normed_offsets = {input_bus: offset}
+                msg = ""
```
Comment (fwitte) [inline src/oemof/solph/components/_offset_converter.py:145]: This is the draft for the compatibility adapter to the old API of the `OffsetConverter`. I think I have to make two more changes (`FutureWarning`/`DepreciationWarning` to be added):

- the coefficients may be iterables, so I would have to calculate the slope and the offset correspondingly
- the max and min may also be iterables, if I see it correctly. What assumptions can be taken here: That the user specifies same length iterables for all of the attributes specified as iterables? May the usage crash without a proper message if he does not?

On top of that, are there more things to be taken into consideration?

Then, there is one last concern: The multi-output `OffsetConverter` made it to the `dev` branch very recently only. Can I consider that feature is not used by anyone yet and simply make it crash, e.g. in the case that users may have specified different `min` or `max` values for every output (which does not make sense, but was possible, see #1067)?

I will also add a test for this section in a different script...

File: src/oemof/solph/components/_offset_converter.py [hunk 3]
```diff
@@ -286,48 +318,59 @@ def _relation_rule(block):
...
+                        block.relation.add(
+                            (n, reference_node[n], f, p, t), (lhs == rhs)
+                        )
 
         self.relation_build = BuildAction(rule=_relation_rule)
+
+
+def calculate_slope_and_offset_with_reference_to_input(
+    max, min, eta_max, eta_min
+):
```
[the remark this answers is not in the document]
Comment (fwitte) [reply src/oemof/solph/components/_offset_converter.py:466]: done

File: docs/usage.rst [hunk 4]
```diff
@@ -688,31 +688,43 @@ The following example illustrates how to define an OffsetConverter for given inf
...
-                min=l_min,
-                nonconvex=solph.NonConvex()
-            ),
-        },
-        coefficients = {bth: (c0, c1)},
-    )
+    >>> diesel_genset = solph.components.OffsetConverter(
+    ...     label='boiler',
+    ...     inputs={bfuel: solph.flows.Flow()},
+    ...     outputs={
```
[the remark this answers is not in the document]
Comment (fwitte) [reply docs/usage.rst:717]: Fixed

File: tests/constraint_tests.py [hunk 5]
```diff
@@ -1356,6 +1356,16 @@ def test_offsetconverter_nonconvex(self):
         b_diesel = solph.buses.Bus(label="bus_diesel")
         b_el = solph.buses.Bus(label="bus_electricity")
 
+        min = 0.2
+        eta_at_nom = 0.4
+        eta_at_min = 0.35
+
+        slope, offset = (
+            solph.components._offset_converter.calculate_slope_and_offset_with_reference_to_output(
```
[the remark this answers is not in the document]
Comment (p-snft) [reply tests/constraint_tests.py:1364]: Independent from the complaint, you should not import from ` _{module}` . If the function is meant to be part of the public API, it should be in the `__init__`. Even if this already solves the line length complaint, you might want to give a shorter function name.

File: src/oemof/solph/components/_offset_converter.py [hunk 6]
```diff
@@ -286,48 +318,59 @@ def _relation_rule(block):
...
+    slope = (max * eta_max - min * eta_min) / (max - min)
+    offset = eta_max - slope
+    return slope, offset
+
+
+def calculate_slope_and_offset_with_reference_to_output(
+    max, min, eta_max, eta_min
+):
```
[the remark this answers is not in the document]
Comment (fwitte) [reply src/oemof/solph/components/_offset_converter.py:507]: added as well

**Comment by fwitte** [comment]
I added two more tests, one for double inputs and double outputs with the `NonConvex` flow at the input, and the same with the `NonConvex` flow at an output.

With that, I think this PR is ready for final review.

**Review by p-snft — CHANGES_REQUESTED** [review]
Thank you for this implementation. It really fills a gap. At the moment, there are few issues left to be fixed before merge.

File: docs/usage.rst [hunk 7]
```diff
@@ -688,37 +688,63 @@ The following example illustrates how to define an OffsetConverter for given inf
...
+    ...         bth: solph.flows.Flow(
+    ...             nominal_value=P_out_max,
+    ...             max=l_max,
+    ...             min=l_min,
+    ...             nonconvex=solph.NonConvex()
+    ...         ),
+    ...     },
+    ...     conversion_factors={bfuel: slope},
+    ...     normed_offsets={bfuel: offset},
```
[the remark this answers is not in the document]
Comment (fwitte) [reply docs/usage.rst:726]: It is a diesel heat boiler, isn't it? The CHP could be in a different example.

File: src/oemof/solph/components/_offset_converter.py [hunk 8]
```diff
@@ -55,14 +57,14 @@ class OffsetConverter(Node):
 
     .. math::
 
-        C_1 = (l_{max}-l_{min})/(l_{max}/\\eta_{max}-l_{min}/\\eta_{min})
+        C_1 = \\frac{(l_{max}/\\eta_{max}-l_{min}/\\eta_{min}}{l_{max}-l_{min}}
```
[the remark this answers is not in the document]
Comment (fwitte) [reply src/oemof/solph/components/_offset_converter.py:60]: Why b? In 5th grade (time runs fast, so might have been 10th grade also) mathematics class it was y=m*x + b for me.

File: src/oemof/solph/components/_offset_converter.py [hunk 9]
```diff
@@ -55,14 +57,14 @@ class OffsetConverter(Node):
...
 
-        C_0 = l_{min} \\cdot (1-C_1/\\eta_{min})
+        C_0 = \\frac{1}{\\eta_{max}} - C_1
```
[the remark this answers is not in the document]
Comment (fwitte) [reply src/oemof/solph/components/_offset_converter.py]: I changed it together with the appearances in other places:

- C_1 -> m
- C_0 -> y_{0,normed}

File: tests/test_components.py [hunk 10]
```diff
@@ -215,39 +215,44 @@ def test_generic_storage_too_many_outputs():
...
+            "`NonConvex` attribute."
+        ),
     ):
         b_el = Bus(label="bus_electricity")
         components.OffsetConverter(
             label="diesel_genset",
             inputs={b_el: Flow()},
             outputs={b_el: Flow()},
-            coefficients={b_el: (2.5, 0.5)},
+            conversion_factors={b_el: (2.5, 0.5)},
```
[the remark this answers is not in the document]
Comment (fwitte) [reply tests/test_components.py:231]: I updated all tests and test names, thank you.

File: src/oemof/solph/components/_offset_converter.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Bachibouzouk) [reply src/oemof/solph/components/_offset_converter.py]: Same for me, `m`is more intuitively a slope variable

File: tests/test_components/test_offset_converter.py
[the remark this answers is not in the document]
Comment (fwitte) [reply tests/test_components/test_offset_converter.py:1]: I added a test (2nd last in this file). It crashes in the `Flow.__init__` actually, before it even reaches the input output relation rules of the `OffsetConverterBlock`. Can we have a quick look together?

File: src/oemof/solph/components/_offset_converter.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (fwitte) [reply src/oemof/solph/components/_offset_converter.py]: I went with `m` and `y_{0, normed}` (https://github.com/oemof/oemof-solph/pull/1068#discussion_r1630676746)

File: docs/usage.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (fwitte) [reply docs/usage.rst]: Adjusted as well

**Comment by fwitte** [comment]
@p-snft, I think, we are finally at it. I implemented a small plotting function for the new figures. I did not reference it anywhere in the docs. Although it is helpful, it could be refined a little bit maybe, I did not want to waste too much time on it.

Have a nice weekend!

**Comment by p-snft** [comment]
The doc error reports that the linked page https://shop.vde.com/en/vde-study-the-cellular-approach has problems. This is no problem to be solved here. There is little to do, yet. I will contribute that myself.

**Review by p-snft — APPROVED** [review]
In my opinion, this is ready to be merged.

There are two remaining problems found by the automated process:
* Docs: Not all linked pages are available. (Not to be solved here.)
* Coverage: Test coverage decreases because of plotting functionality. As, however, this is not an advertised feature and nobody will rely on it, I think this is in order.

**Pull request merged**